### PR TITLE
Some fixes to make JSON properly recognize numbers in quotes.

### PR DESCRIPTION
**Pull request by haberman**

### CODE REVIEW

File: upb/json/parser.c [hunk 1]
```diff
@@ -608,103 +609,156 @@ static void start_number(upb_json_parser *p, const char *ptr) {
...
-      if (errno == ERANGE || end != myend)
-        goto err;
-      else
-        upb_sink_putdouble(&p->top->sink, parser_getsel(p), val);
+    default:
       break;
+  }
+
+  if (type != UPB_TYPE_DOUBLE && type != UPB_TYPE_FLOAT && is_quoted) {
+    /* Quoted numbers shouldn't support double forms for integer types. */
```
Comment (scwhittle) [inline upb/json/parser.c]: I think this is a little clearer
/* Quoted numbers for integer types are not allowed to be in double form */

File: upb/json/parser.c [hunk 2]
```diff
@@ -608,103 +609,156 @@ static void start_number(upb_json_parser *p, const char *ptr) {
...
-      break;
+  }
+
+  switch (type) {
+#define CASE(capitaltype, smalltype, min, max)                            \
+    case UPB_TYPE_ ## capitaltype: {                                      \
+      if (modf(val, &dummy) != 0 || val > max || val < min) {             \
+        return false;                                                     \
+      } else {                                                            \
+        upb_sink_put ## smalltype(&p->top->sink, parser_getsel(p), val);  \
```
Comment (scwhittle) [inline upb/json/parser.c]: cast val to smalltype for clarity?

File: upb/json/parser.c [hunk 3]
```diff
@@ -608,103 +609,156 @@ static void start_number(upb_json_parser *p, const char *ptr) {
...
-  return parse_number(p);
+  return parse_number(p, false);
 }
 
-static bool parse_number(upb_json_parser *p) {
-  size_t len;
-  const char *buf;
-  const char *myend;
+static bool parse_number_from_buffer(upb_json_parser *p, const char *buf,
+                                     const char *bufend, bool is_quoted) {
```
Comment (scwhittle) [inline upb/json/parser.c]: I was somewhat confused at first thinking is_quoted meant that buf included quotes, not that buf is located inside of quotes. comment?

File: upb/json/parser.c [hunk 4]
```diff
@@ -608,103 +609,156 @@ static void start_number(upb_json_parser *p, const char *ptr) {
...
+    CASE(INT64, int64, INT64_MIN, INT64_MAX);
+    CASE(UINT32, uint32, 0, UINT32_MAX);
+    CASE(UINT64, uint64, 0, UINT64_MAX);
+#undef CASE
+
+    case UPB_TYPE_DOUBLE:
+      upb_sink_putdouble(&p->top->sink, parser_getsel(p), val);
+      return true;
+    case UPB_TYPE_FLOAT:
+      if (false /*val > FLT_MAX || val < -FLT_MAX*/) {
```
Comment (scwhittle) [inline upb/json/parser.c]: fix?

File: upb/json/parser.c (hunk 3 above)
Comment (haberman) [inline upb/json/parser.c]: Done.

File: upb/json/parser.c [hunk 5]
```diff
@@ -608,103 +609,156 @@ static void start_number(upb_json_parser *p, const char *ptr) {
...
+  size_t len = bufend - buf;
   char *end;
+  upb_fieldtype_t type = upb_fielddef_type(p->top->f);
+  double val;
+  double dummy;
 
-  /* strtol() and friends unfortunately do not support specifying the length of
-   * the input string, so we need to force a copy into a NULL-terminated buffer. */
-  if (!multipart_text(p, "\0", 1, false)) {
+  if (buf[0] == ' ') {
```
Comment (haberman) [inline upb/json/parser.c]: Done.

File: upb/json/parser.c [hunk 6]
```diff
@@ -608,103 +609,156 @@ static void start_number(upb_json_parser *p, const char *ptr) {
...
+   * strtod().
+   */
+  switch (type) {
     case UPB_TYPE_ENUM:
     case UPB_TYPE_INT32: {
-      long val = strtol(p->accumulated, &end, 0);
-      if (val > INT32_MAX || val < INT32_MIN || errno == ERANGE || end != myend)
-        goto err;
-      else
+      long val = strtol(buf, &end, 0);
```
Comment (haberman) [inline upb/json/parser.c]: Done.

File: upb/json/parser.c (hunk 1 above)
Comment (haberman) [inline upb/json/parser.c]: Done.

File: upb/json/parser.c (hunk 2 above)
Comment (haberman) [inline upb/json/parser.c]: Done.

File: upb/json/parser.c (hunk 4 above)
Comment (haberman) [inline upb/json/parser.c]: Done.

**Comment by haberman** [comment]
Responded to comments, PTAL.

**Comment by scwhittle** [comment]
LGTM

**Pull request merged**